scipy/special/ellint_carlson_cpp_lite/_rd.hh

```diff
@@ -0,0 +1,158 @@
+    }
+#endif
+
+    if ( !( argcheck::ph_good(x) && argcheck::ph_good(y) &&
+            argcheck::ph_good(z) ) )
+    {
+	res = typing::nan<T>();
+	return ExitStatus::bad_args;
+    }
+
+    if ( argcheck::too_small(z) )
+    {
+	res = typing::huge<T>();
+	return ExitStatus::singular;
+    }
+
+    if ( argcheck::isinf(x) || argcheck::isinf(y) || argcheck::isinf(z) )
+    {
+	res = T(0.0);
+	return status;
+    }
+
+    if ( argcheck::too_small(x) && argcheck::too_small(y) )
+    {
+	res = typing::huge<T>();
+	return ExitStatus::singular;
+    }
+
+    cct1[0] = x;
+    cct1[1] = y;
+    cct1[2] = z;
+    cct1[3] = z;
+    cct1[4] = z;
+    T Am = arithmetic::nsum2(cct1, 5) / (RT)5.0;
+    T xm = x;
+    T ym = y;
+    T zm = z;
+    T xxm = Am - xm;
+    T yym = Am - ym;
+    RT fterm = std::max({std::abs(xxm), std::abs(yym), std::abs(Am - z)}) /
+               arithmetic::ocrt(rerr / 5.0);
+    RT d4m(1.0);
+    T adt(0.0), ade(0.0);
+
+    unsigned int m = 0;
+    RT aAm;
+    T tmp;
+    while ( (aAm = std::abs(Am)) <= fterm ||
+	    aAm <= std::max({std::abs(xxm), std::abs(yym), std::abs(Am - zm)}) )
+    {
+	if ( m > config::max_iter )
+	{
+	    status = ExitStatus::n_iter;
+	    break;
+	}
+	cct1[0] = cct2[2] = std::sqrt(xm);
+	cct1[1] = cct2[0] = std::sqrt(ym);
+	cct1[2] = cct2[1] = std::sqrt(zm);
+	T lam = arithmetic::ndot2(cct1, cct2, 3);
+
+	/* Ref[2], Eq. 2.34, accumulating the summation term into adt
+	 * with compensation term ade */
+	tmp = d4m / (cct1[2] * (zm + lam));
+	arithmetic::sum2_acc(tmp, adt, ade);
+
+        Am = (Am + lam) * (RT)0.25;
+        xm = (xm + lam) * (RT)0.25;
+        ym = (ym + lam) * (RT)0.25;
+        zm = (zm + lam) * (RT)0.25;
+        xxm *= (RT)0.25;
+        yym *= (RT)0.25;
+        fterm *= (RT)0.25;
+        d4m *= (RT)0.25;
+
+        ++m;
+    }
+    /* Burn some extra cycles re-balancing Am as the "true" centroid */
+    cct1[0] = xm;
+    cct1[1] = ym;
+    cct1[2] = zm;
+    cct1[3] = zm;
+    cct1[4] = zm;
+    Am = arithmetic::nsum2(cct1, 5) / (RT)5.0;
+    xxm /= Am;
+    yym /= Am;
+    T zzm = (xxm + yym) / (RT)(-3.0);
+    T xy = xxm * yym;
+    T zz2 = zzm * zzm;
+    T e2 = xy - zz2 * (RT)6.0;
+    T e3 = (xy * (RT)3.0 - zz2 * (RT)8.0) * zzm;
+    T e4 = (xy - zz2) * zz2 * (RT)3.0;
+    T e5 = xy * zz2 * zzm;
+    T t = std::sqrt(Am);
+    tmp = d4m / (t * t * t);
+    cct1[0] = arithmetic::comp_horner(e2, constants::RDJ_C1);
+    cct1[1] = arithmetic::comp_horner(e3, constants::RDJ_C2);
+    cct1[2] = arithmetic::comp_horner(e2, constants::RDJ_C3);
+    cct1[3] = arithmetic::comp_horner(e2, constants::RDJ_C4);
+    cct1[4] = arithmetic::comp_horner(e2, constants::RDJ_C5);
+    cct1[5] = e3 * (RT)(constants::RDJ_C5[1]);
+
+    cct2[0] = T(1.0);
+    cct2[1] = T(1.0);
+    cct2[2] = e3;
+    cct2[3] = e4;
+    cct2[4] = e5;
+    cct2[5] = e4;
+    t = arithmetic::dot2(cct1, cct2) / (RT)(constants::RDJ_DENOM) + 1.0;
+    tmp *= t;
+    tmp += adt * (RT)3.0;
+
```

Comment:
This is certainly not clear, but I assume that the arithmetic involved in computing `t` from `cct1` and `cct2` does implement (41). For example, looking at the coefficient of the linear term for e2 (in `cct1[0]`, and using `constants::RDJ_C1[1]` divided by `constants::RDJ_DENOM`) is indeed -3/14. I have not verified the others.

Do you think that comments are needed to explain what the code is doing?

Reply:
Thank for pointing this out. I wasn't clear in that the formula used here (and in the computation of RJ) is the higher-order one from [the DLMF](https://dlmf.nist.gov/19.36#E2); there are more terms than Equations (32) and (41) in Carlson (1994). This will be addressed by code comments.

In addition, the coefficients in those formulas in the paper and manual were written as fractions. However, most of those values were not exactly representable as floating-point. To work around this, I converted the fractions to a common denominator `constants::RDJ_DENOM` by which the value is then divided.

Reply:
Gotcha! That makes sense. Thanks for explaining and thanks for your work on this.

Reply:
Comments added in 0510b860 and 0c40e91b0